### PR TITLE
feat: TypeSafeHTTPBasic

### PR DESCRIPTION
**Pull request by Andrew-Lees11**
Once [typeSafeCredentials](https://github.com/IBM-Swift/Kitura-Credentials/pull/64) is merged we can implement HTTP basic auth using typeSafeMiddlware. 

Docs description:
A `TypeSafeCredentials` plugin for HTTP basic authentication. This protocol will be implemented by a Swift object defined by the user. The plugin must implement a `verifyPassword` function which takes a username and password as input and returns an instance of `Self` on success or `nil` on failure. This instance must contain the authentication `provider` (defaults to "HTTPBasic") and an `id`, uniquely identifying the user. The users object can then be used in TypeSafeMiddlware routes to authenticate with HTTP basic.

### CODE REVIEW

**Comment by Andrew-Lees11** [comment]
I have removed inProgress from authenticate since in current credentials it will never be invoked. This will make it different from credentials authenticate signature so i was unsure if we want to maintain consistency?

**Comment by Andrew-Lees11** [comment]
The tests are currently failing since they require typeSafeMiddleware which will only be available in Kitura 2.4. and it will need TypeSafeCredentials to be tagged.

File: README.md [hunk 1]
```diff
@@ -56,6 +56,42 @@ CredentialsHTTPDigest initialization is similar to CredentialsHTTPBasic. In addi
...
+adding any instance variables which you will initialise in verifyPassword:
+
+```swift
+import CredentialsHTTP
+
+public struct MyBasicAuth: TypeSafeHTTPBasic {
+
+    public let id: String
+
+    public static let users = ["John" : "12345", "Mary" : "qwerasdf"]
```
Comment (ianpartridge) [inline README.md]: Probably make this private!

**Comment by djones6** [comment]
@Andrew-Lees11 FYI, Kitura and Credentials dependencies are merged and tagged - I have restarted the CI.

File: Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import KituraNet
+import Credentials
+
+@testable import CredentialsHTTP
+
+class TestTypeSafeBasic : XCTestCase {
+    
+    static var allTests : [(String, (TestTypeSafeBasic) -> () throws -> Void)] {
+        return [
+            ("testTypeSafeNoCredentials", testTypeSafeNoCredentials),
```
Comment (ianpartridge) [inline Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift:30]: This array is missing some of the tests!

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    ///                     authentication data in the request.
+    public static func authenticate(request: RouterRequest, response: RouterResponse, onSuccess: @escaping (Self) -> Void, onFailure: @escaping (HTTPStatusCode?, [String : String]?) -> Void, onSkip: @escaping (HTTPStatusCode?, [String : String]?) -> Void) {
+        
+        let userid: String
+        let password: String
+        if let requestUser = request.urlURL.user, let requestPassword = request.urlURL.password {
+            userid = requestUser
+            password = requestPassword
+        }
+        else {
```
Comment (ianpartridge) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: Please put the else on the previous line.

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            userid = credentials[0]
+            password = credentials[1]
+        }
+        
+        verifyPassword(userid, password) { selfInstance in
+            if let selfInstance = selfInstance {
+                onSuccess(selfInstance)
+            }
+            else {
```
Comment (ianpartridge) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: Same comment.

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            let authorizationHeaderComponents = authorizationHeader.components(separatedBy: " ")
+            guard authorizationHeaderComponents.count == 2,
+                authorizationHeaderComponents[0] == "Basic",
+                let decodedData = Data(base64Encoded: authorizationHeaderComponents[1], options: Data.Base64DecodingOptions(rawValue: 0)),
+                let userAuthorization = String(data: decodedData, encoding: .utf8) else {
+                    return onSkip(.unauthorized, ["WWW-Authenticate" : "Basic realm=\"" + realm + "\""])
+            }
+            let credentials = userAuthorization.components(separatedBy: ":")
+            guard credentials.count >= 2 else {
+                onFailure(.badRequest, nil)
```
Comment (ianpartridge) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: `return onFailure...`

File: Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        performServerTest(router: router) { expectation in
+            self.performRequest(method: "get", path:"/private/typesafebasic", callback: {response in
+                XCTAssertNotNil(response, "ERROR!!! ClientRequest response object was nil")
+                XCTAssertEqual(response?.statusCode, HTTPStatusCode.OK, "HTTP Status code was \(String(describing: response?.statusCode))")
+                do {
+                    let body = try response?.readString()
+                    XCTAssertEqual(body,"{\"name\":\"Mary\",\"provider\":\"HTTPBasic\"}")
+                }
+                catch{
```
Comment (ianpartridge) [inline Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift]: Merge with previous line.

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ }
+ ```
+ */
+public protocol TypeSafeHTTPBasic : TypeSafeCredentials {
+    
+    /// The realm for which these credentials are valid (defaults to "User")
+    static var realm: String { get }
+    
+    /// The closure which takes a username and password and returns a TypeSafeHTTPBasic instance on success or nil on failure.
+    static var verifyPassword: ((String, String, @escaping (Self?) -> Void) -> Void) { get }
```
Comment (ianpartridge) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: This isn't very type-safe, did you consider replacing these Strings with something protocol-based?

**Review by ianpartridge — CHANGES_REQUESTED** [review]

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift (hunk 7 above)
Comment (djones6) [reply Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: I think we should replace with something like
```swift
static func verifyPassword(username: String, password, String, callback: @escaping (Self?) -> Void) -> Void
```
Xcode will auto-complete this much nicer than the version where we define a closure signature, and apply the labels which indicates how to handle each argument.

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ ```swift
+ public struct MyHTTPBasic: TypeSafeHTTPBasic {
+ 
+    public var id: String
+ 
+    static let users = ["John" : "12345", "Mary" : "qwerasdf"]
+ 
+    public static let realm = "Login message"
+ 
+    public static var verifyPassword: ((String, String, @escaping (MyHTTPBasic?) -> Void) -> Void) =
```
Comment (djones6) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: Usage example needs updating now that this is a function

File: Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    respondWith(user, nil)
+ }
+ ```
+ */
+public protocol TypeSafeHTTPBasic : TypeSafeCredentials {
+    
+    /// The realm for which these credentials are valid (defaults to "User")
+    static var realm: String { get }
+    
+    /// The closure which takes a username and password and returns a TypeSafeHTTPBasic instance on success or nil on failure.
```
Comment (djones6) [inline Sources/CredentialsHTTP/TypeSafeHTTPBasic.swift]: Comment needs updating now that this is a function

**Review by djones6 — CHANGES_REQUESTED** [review]

File: Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift (hunk 2 above)
Comment (Andrew-Lees11) [reply Tests/CredentialsHTTPTests/TestTypeSafeBasic.swift:30]: Tests have been added and this file has been added to linux main

**Review by djones6 — APPROVED** [review]

**Review by ianpartridge — APPROVED** [review]

**Pull request merged**